### PR TITLE
DCS-997 user input trimmed to remove whitespace

### PR DESCRIPTION
**Pull request by GurnankCheema**
The prisonNumber, firstName and lastName have had any whitespaces trimmed. This will overcome any accidental input of whitespace for example if the user copy-pastes values from elsewhere.

### CODE REVIEW

File: backend/routes/createBooking/prisonerSearch/prisonerSearchController.js [hunk 1]
```diff
@@ -19,9 +19,9 @@ module.exports =
       searchResults = await prisonApi.globalSearch(
         res.locals,
         {
-          offenderNo: prisonNumber,
-          lastName,
-          firstName,
+          offenderNo: prisonNumber && prisonNumber.trim(),
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchController.js:22]: Can use optional chaining for these:
`prisonNumber && prisonNumber.trim()` => `prisonNumber.trim?.()`

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.js [hunk 2]
```diff
@@ -1,25 +1,25 @@
 const dobValidation = require('../../../shared/dobValidation')
 
 module.exports = formValues => {
-  const { firstName, lastName, prisonNumber, dobDay, dobMonth, dobYear } = formValues
+  const { firstName = '', lastName = '', prisonNumber = '', dobDay, dobMonth, dobYear } = formValues
   const { dobErrors } = dobValidation(dobDay, dobMonth, dobYear)
   const videolinkPrisonerSearchErrors = [...dobErrors]
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.js:6]: Don't really want to be trimming and re-trimming all over the shop. 

This adds a whole extra unnecessary dimension of complexity throughout the rest of the code. 

rather than: 
```
const { firstName = '', lastName = '', prisonNumber = '', dobDay, dobMonth, dobYear } = formValues
```

You could do something like
```
const { firstName: firstNameInput, lastName: lastNameInput, prisonNumber: prisonNumberInput , dobDay, dobMonth, dobYear } = formValues

const firstName = firstNameInput.trim?.()
const lastName = lastNameNameInput.trim?.()
const prisonNumber = prisonNumberInput.trim?.()
```

And then leave the rest of the code the same?

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.test.js [hunk 3]
```diff
@@ -19,6 +19,18 @@ describe('prisoner search validation', () => {
     ])
   })
 
+  it('should error if prison number has 7 charcters including some are whitespaces ', async () => {
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.test.js:22]: Add tests for when people search for whitespace only characters?

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.js (hunk 2 above)
Comment (GurnankCheema) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.js:6]: done

File: backend/routes/createBooking/prisonerSearch/prisonerSearchController.js (hunk 1 above)
Comment (GurnankCheema) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchController.js:22]: done

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.test.js (hunk 3 above)
Comment (GurnankCheema) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.test.js]: done

File: backend/utils/nunjucksSetup.js [hunk 4]
```diff
@@ -181,6 +181,10 @@ module.exports = (app, path) => {
     return value || specifiedText || '--'
   })
 
+  njkEnv.addFilter('trimWhitespace', text => {
```
Comment (GurnankCheema) [inline backend/utils/nunjucksSetup.js:184]: created a filter rather than trimming whitespace in the data returned by the controller because the filter will be usable elsewhere.

File: backend/api/prisonApi.ts [hunk 5]
```diff
@@ -8,10 +8,10 @@ type GlobalSearchRequest = {
   offenderNo: string
   lastName: string
   firstName: string
-  gender: 'F' | 'M' | 'NK' | 'NS'
+  gender?: 'F' | 'M' | 'NK' | 'NS'
   location: 'IN' | 'OUT' | 'ALL'
   dateOfBirth: string
-  includeAliases: boolean
+  includeAliases?: boolean
```
Comment (andrewrlee) [inline backend/api/prisonApi.ts:14]: these aren't optional are they? If doing for tests you can just `as GlobalSearchRequest`

File: backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts [hunk 6]
```diff
@@ -84,21 +79,21 @@ describe('Video link prisoner search', () => {
               latestLocationId: 'MDI',
               latestLocation: 'Moorlands',
             },
-          ])
+          ] as never)
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts:82]: shouldn't this be `PrisonerDetail[]`, casting something to `never` probably doesn't make sense conceptually

Comment (andrewrlee) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts:82]: you will need to `mockResolvedValue`, maybe that was causing issues?

File: backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts [hunk 7]
```diff
@@ -1,24 +1,18 @@
-const prisonerSearchController = require('./prisonerSearchController')
-const config = require('../../../config')
+import PrisonerSearchController from './prisonerSearchController'
+import { mockRequest, mockResponse } from '../../__test/requestTestUtils'
+import config from '../../../config'
+import PrisonApi from '../../../api/prisonApi'
 
+const prisonApi = new PrisonApi(null) as jest.Mocked<PrisonApi>
+jest.mock('../../../api/prisonApi')
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts:7]: probably should swap these lines, the convention we have is generally:
1.)  imports,
2.) mocking
3.) declarations.

File: backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts [hunk 8]
```diff
@@ -206,16 +203,14 @@ describe('Video link prisoner search', () => {
 
     describe('when there are API errors', () => {
       it('should render the error template if there is an error retrieving agencies', async () => {
-        oauthApi.userRoles.mockReturnValue([{ roleCode: 'VIDEO_LINK_COURT_USER' }])
-        prisonApi.getAgencies.mockImplementation(() => Promise.reject(new Error('Network error')))
-        await expect(controller(req, res)).rejects.toThrow('Network error')
+        prisonApi.getAgencies.mockRejectedValue(new Error('Network error'))
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts:206]: nice 👍

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+type FormValues = {
+  firstName?: string
+  lastName?: string
+  prisonNumber?: string
+  dobDay?: string
+  dobMonth?: string
+  dobYear?: string
+}
+export default class PrisonerSearchValidation {
+  public validate(formValues: FormValues): ValidationError[] {
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts:14]: A (possibly better) alternative would be to use the type guard we have for `assertHasOptionalStringValues`

```
public validate(form: unknown): ValidationError[] {
     assertHasOptionalStringValues(form, [
    'firstName',
    'lastName',
    'prisonNumber',
    'dobDay',
    'dobMonth',
    'dobYear',
  ]) 
  // can now refer to all the field, e.g: form.firstName etc... as a nullable string
```

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    const { firstName, lastName, prisonNumber, dobDay, dobMonth, dobYear } = formValues
+    const { dobErrors } = dobValidation(dobDay, dobMonth, dobYear)
+    const videolinkPrisonerSearchErrors = [...dobErrors]
+
```
Comment (andrewrlee) [inline backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts:18]: we're still doing the trimming in loads of places, would reassign them and do them in one place rather than through out the code here..

File: views/createBooking/prisonerSearch.njk [hunk 11]
```diff
@@ -6,72 +6,61 @@
...
 {% set title = 'Search for a prisoner' %}
-
 {% macro errorClass(field) -%}
   {{ ' govuk-input--error' if errors | hasErrorWithPrefix([field, 'dobError']) else '' }}
 {%- endmacro %}
-
 {% block content %}
-  {% if errors.length > 0 %}
-    {{ govukErrorSummary({
+{% if errors.length > 0 %}
```
Comment (andrewrlee) [inline views/createBooking/prisonerSearch.njk:14]: your formatting seems too far over to the left here..

Comment (andrewrlee) [reply views/createBooking/prisonerSearch.njk:14]: the block/if etc.. should all indent the code in

File: backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts (hunk 6 above)
Comment (GurnankCheema) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts:82]: yes. changed

File: backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts (hunk 7 above)
Comment (GurnankCheema) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchController.test.ts:7]: changed order

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts (hunk 10 above)
Comment (GurnankCheema) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts:18]: trimming now done before validation stage

File: views/createBooking/prisonerSearch.njk (hunk 11 above)
Comment (GurnankCheema) [reply views/createBooking/prisonerSearch.njk:14]: macro removed as no longer needed

File: backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts (hunk 9 above)
Comment (GurnankCheema) [reply backend/routes/createBooking/prisonerSearch/prisonerSearchValidation.ts:14]: assertHasOptionalStringValues used

File: views/createBooking/prisonerSearch.njk (hunk 11 above)
Comment (andrewrlee) [reply views/createBooking/prisonerSearch.njk:14]: your inside the content block - all the content in the block should be indented

**Review by andrewrlee — APPROVED** [review]
LGTM
